### Simple Exponential Smoothing on Home Sales

In [1]:
# setup absolute path to location of package Starts and config-file 
from inspect import getsourcefile
import os.path as path, sys
current_dir = path.dirname(path.abspath(getsourcefile(lambda:0)))
sys.path.insert(0, current_dir[:current_dir.rfind(path.sep)])

from Starts.startml import *
from Starts.startvis import *
from Starts.startspk import *
%matplotlib inline

local_kwargs {'data_path': './data/single-family-home-sales.xlsx, ./data/champagne-sales.xlsx', 'drop_obj_col': False, 'nan_drop_col': False, 'nan_drop_row': False, 'nan_zero': False, 'nan_mean': True, 'nan_mean_neighbors': False}


In [2]:
sales = idata[0]
sales.head()

Month  Home Sales  Alpha .8  Forecast  Alpha .6  Forecast.1  Alpha .2  \
0  1990-01        45.0       NaN       NaN       NaN         NaN       NaN   
1  1990-02        50.0       NaN       NaN       NaN         NaN       NaN   
2  1990-03        58.0       NaN       NaN       NaN         NaN       NaN   
3  1990-04        52.0       NaN       NaN       NaN         NaN       NaN   
4  1990-05        50.0       NaN       NaN       NaN         NaN       NaN   

   Forecast.2  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN

In [3]:
sales.tail()

Month  Home Sales  Alpha .8  Forecast  Alpha .6  Forecast.1  Alpha .2  \
67  1995-08        63.0       NaN       NaN       NaN         NaN       NaN   
68  1995-09        55.0       NaN       NaN       NaN         NaN       NaN   
69  1995-10        54.0       NaN       NaN       NaN         NaN       NaN   
70  1995-11        44.0       NaN       NaN       NaN         NaN       NaN   
71  1995-12         NaN       NaN       NaN       NaN         NaN       NaN   

    Forecast.2  
67         NaN  
68         NaN  
69         NaN  
70         NaN  
71         NaN

In [4]:
StartML.convert_time_series(sales, time_column='Month', form='%Y-%m').head()

Home Sales  Alpha .8  Forecast  Alpha .6  Forecast.1  Alpha .2  \
Month                                                                        
1990-01-01        45.0       NaN       NaN       NaN         NaN       NaN   
1990-02-01        50.0       NaN       NaN       NaN         NaN       NaN   
1990-03-01        58.0       NaN       NaN       NaN         NaN       NaN   
1990-04-01        52.0       NaN       NaN       NaN         NaN       NaN   
1990-05-01        50.0       NaN       NaN       NaN         NaN       NaN   

            Forecast.2  
Month                   
1990-01-01         NaN  
1990-02-01         NaN  
1990-03-01         NaN  
1990-04-01         NaN  
1990-05-01         NaN

In [5]:
hsales = sales['Home Sales']

In [6]:
# Calculate weight with different smooting parameters alpha = 0.8, 0.6, 0.2
val_hsales = hsales.values[:-1]
weight_alpha_8 = [0.8*(1-0.8)**i for i in reversed(range(len(val_hsales)))]
weight_alpha_6 = [0.6*(1-0.6)**i for i in reversed(range(len(val_hsales)))]
weight_alpha_2 = [0.2*(1-0.2)**i for i in reversed(range(len(val_hsales)))]

In [7]:
sales.columns, len(sales['Alpha .8'])

(Index(['Home Sales', 'Alpha .8', 'Forecast', 'Alpha .6', 'Forecast.1',
        'Alpha .2', 'Forecast.2'],
       dtype='object'), 72)

In [8]:
sales['Alpha .8'][:-1] = weight_alpha_8
sales['Alpha .6'][:-1] = weight_alpha_6
sales['Alpha .2'][:-1] = weight_alpha_2

In [9]:
sales.head()

Home Sales      Alpha .8  Forecast      Alpha .6  Forecast.1  \
Month                                                                      
1990-01-01        45.0  9.444733e-50       NaN  8.362779e-29         NaN   
1990-02-01        50.0  4.722366e-49       NaN  2.090695e-28         NaN   
1990-03-01        58.0  2.361183e-48       NaN  5.226737e-28         NaN   
1990-04-01        52.0  1.180592e-47       NaN  1.306684e-27         NaN   
1990-05-01        50.0  5.902958e-47       NaN  3.266711e-27         NaN   

                Alpha .2  Forecast.2  
Month                                 
1990-01-01  3.291009e-08         NaN  
1990-02-01  4.113761e-08         NaN  
1990-03-01  5.142202e-08         NaN  
1990-04-01  6.427752e-08         NaN  
1990-05-01  8.034690e-08         NaN

In [10]:
forecast = [sales['Home Sales'].values[i]*sales['Alpha .8'].values[i] for i in reversed(range(len(val_hsales)))]
forecast_1 = [sales['Home Sales'].values[i]*sales['Alpha .6'].values[i] for i in reversed(range(len(val_hsales)))]
forecast_2 = [sales['Home Sales'].values[i]*sales['Alpha .2'].values[i] for i in reversed(range(len(val_hsales)))]
sales['Forecast'][:-1] = forecast
sales['Forecast.1'][:-1] = forecast_1
sales['Forecast.2'][:-1] = forecast_2
sales.head()

Home Sales      Alpha .8  Forecast      Alpha .6  Forecast.1  \
Month                                                                      
1990-01-01        45.0  9.444733e-50  35.20000  8.362779e-29    26.40000   
1990-02-01        50.0  4.722366e-49   8.64000  2.090695e-28    12.96000   
1990-03-01        58.0  2.361183e-48   1.76000  5.226737e-28     5.28000   
1990-04-01        52.0  1.180592e-47   0.40320  1.306684e-27     2.41920   
1990-05-01        50.0  5.902958e-47   0.08192  3.266711e-27     0.98304   

                Alpha .2  Forecast.2  
Month                                 
1990-01-01  3.291009e-08     8.80000  
1990-02-01  4.113761e-08     8.64000  
1990-03-01  5.142202e-08     7.04000  
1990-04-01  6.427752e-08     6.45120  
1990-05-01  8.034690e-08     5.24288

In [11]:
sales.tail()

Home Sales  Alpha .8      Forecast  Alpha .6    Forecast.1  \
Month                                                                    
1995-08-01        63.0    0.0064  6.139076e-46    0.0384  6.794758e-26   
1995-09-01        55.0    0.0320  1.369486e-46    0.0960  3.031508e-26   
1995-10-01        54.0    0.1600  2.361183e-47    0.2400  1.045347e-26   
1995-11-01        44.0    0.8000  4.250130e-48    0.6000  3.763251e-27   
1995-12-01         NaN       NaN           NaN       NaN           NaN   

            Alpha .2  Forecast.2  
Month                             
1995-08-01    0.1024    0.000003  
1995-09-01    0.1280    0.000003  
1995-10-01    0.1600    0.000002  
1995-11-01    0.2000    0.000001  
1995-12-01       NaN         NaN

In [12]:
forecast_dec = np.sum(sales['Forecast'].values[:-1]) 
forecast_1_dec = np.sum(sales['Forecast.1'].values[:-1]) 
forecast_2_dec = np.sum(sales['Forecast.2'].values[:-1]) 

forecast_dec, forecast_1_dec, forecast_2_dec

(46.105470350583452, 48.683081186078937, 54.838020063318503)

In [13]:
sales.loc['1995-12','Forecast']= forecast_dec
sales.loc['1995-12','Forecast.1']= forecast_1_dec
sales.loc['1995-12','Forecast.2']= forecast_2_dec

In [14]:
sales.fillna(0, inplace=True)
sales.tail()

Home Sales  Alpha .8      Forecast  Alpha .6    Forecast.1  \
Month                                                                    
1995-08-01        63.0    0.0064  6.139076e-46    0.0384  6.794758e-26   
1995-09-01        55.0    0.0320  1.369486e-46    0.0960  3.031508e-26   
1995-10-01        54.0    0.1600  2.361183e-47    0.2400  1.045347e-26   
1995-11-01        44.0    0.8000  4.250130e-48    0.6000  3.763251e-27   
1995-12-01         0.0    0.0000  4.610547e+01    0.0000  4.868308e+01   

            Alpha .2  Forecast.2  
Month                             
1995-08-01    0.1024    0.000003  
1995-09-01    0.1280    0.000003  
1995-10-01    0.1600    0.000002  
1995-11-01    0.2000    0.000001  
1995-12-01    0.0000   54.838020

#### The result of forecast on 1995-12 is calculated with different alpha as following:
#### Alpha = 0.8, forecast = 46.10547
#### Alpha = 0.6, forecast = 48.68308
#### Alpha = 0.2, forecast = 54.838020